### ISSUE TITLE
ALE doesn't lint some errors gcc spits out

### ISSUE THREAD

**Issue by Kimplul** [dense-analysis/ale#3320]
## Information

**VIM version**
NVIM v0.4.4
Build type: Release

Operating System: Debian 10

## What went wrong
ALE doesn't lint certain errors, such as the one that occurs in this specific situation:
c.cpp:
```
#include <string>

int main(){
    std::string str = TEST;
    return 0;
}
```

Output of :ALEInfo:
```
(executable check - success) g++
(finished - exit code 1) ['/bin/bash', '-c', '''g++'' -S -x c++ -o /dev/null -iquote ''/home/ki
mi/Documents/Projects/minimalale'' ''-DTEST=''\''''h''\'''''' -std=c++14 -Wall - < ''/tmp/nvim8
0dfds/2/c.cpp''']
<<<OUTPUT STARTS>>>
<stdin>: In function ‘int main()’:
<command-line>: error: conversion from ‘char’ to non-scalar type ‘std::string’ {aka ‘std::__cxx
11::basic_string<char>’} requested
<stdin>:4:23: note: in expansion of macro ‘TEST’
<<<OUTPUT ENDS>>>
```

To clear it up a bit, I'm using a compile-time macro to assign a single character to a string. This is an error, as they're not the same type, but because of the somewhat odd circumstances of the error message, ale skips over it.

I found a quick botch. I modified line 12 in ```autoload/ale/handlers/gcc.vim``` to be:
```
let s:pattern = '\v^([a-zA-Z]?:?[^:]+):(\d+)?:?(\d+)?:? ([^:]+): (.+)$'
```

This fix passes all test cases, but again, don't know if it has any weird side-effects that the tests don't pick up on.

## Reproducing the bug
1. Create example file
2. Create compile_commands.json file with a macro that corresponds to a single character, in my case
```
[
    {
        "arguments": [
            "g++",
            "-c",
            "-DTEST='h'",
            "c.cpp"
        ],
        "directory": "/home/kimi/Documents/Projects/minimalale",
        "file": "c.cpp"
    }
]
```

### :ALEInfo
```
 Current Filetype: cpp
Available Linters: ['cc', 'ccls', 'clangcheck', 'clangd', 'clangtidy', 'clazy', 'cppcheck', 'cpplint', 'cquery', 'flawfinder']
   Linter Aliases:
'cc' -> ['gcc', 'clang', 'g++', 'clang++']
  Enabled Linters: ['cc', 'ccls', 'clangcheck', 'clangd', 'clangtidy', 'clazy', 'cppcheck', 'cpplint', 'cquery', 'flawfinder']
 Suggested Fixers: 
  'astyle' - Fix C/C++ with astyle.
  'clang-format' - Fix C/C++ and cuda files with clang-format.
  'clangtidy' - Fix C/C++ and ObjectiveC files with clang-tidy.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'uncrustify' - Fix C, C++, C#, ObjectiveC, ObjectiveC++, D, Java, Pawn, and VALA files with uncrustify.
 Linter Variables:

let g:ale_cpp_cc_executable = 'g++'
let g:ale_cpp_cc_options = '-std=c++14 -Wall'
let g:ale_cpp_ccls_executable = 'ccls'
let g:ale_cpp_ccls_init_options = {}
let g:ale_cpp_clangcheck_executable = 'clang-check'
let g:ale_cpp_clangcheck_options = ''
let g:ale_cpp_clangd_executable = 'clangd'
let g:ale_cpp_clangd_options = ''
let g:ale_cpp_clangtidy_checks = []
let g:ale_cpp_clangtidy_executable = 'clang-tidy'
let g:ale_cpp_clangtidy_extra_options = ''
let g:ale_cpp_clangtidy_options = ''
let g:ale_cpp_clazy_checks = ['level1']
let g:ale_cpp_clazy_executable = 'clazy-standalone'
let g:ale_cpp_clazy_options = ''
let g:ale_cpp_cppcheck_executable = 'cppcheck'
let g:ale_cpp_cppcheck_options = '--enable=style'
let g:ale_cpp_cpplint_executable = 'cpplint'
let g:ale_cpp_cpplint_options = ''
let g:ale_cpp_cquery_cache_directory = '/home/kimi/.cache/cquery'
let g:ale_cpp_cquery_executable = 'cquery'
let g:ale_cpp_flawfinder_executable = 'flawfinder'
let g:ale_cpp_flawfinder_minlevel = 1
let g:ale_cpp_flawfinder_options = ''
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'*': ['remove_trailing_lines', 'trim_whitespace'], 'python': ['autopep8']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) g++
(finished - exit code 1) ['/bin/bash', '-c', '''g++'' -S -x c++ -o /dev/null -iquote ''/home/kimi/Documents/Projects/minimalale'' ''-DTEST=''\''''h''\'''''' -std=c++14 -Wall - < ''/tmp/nvimjP5eNw/2/c.cpp''']

<<<OUTPUT STARTS>>>
<stdin>: In function ‘int main()’:
<command-line>: error: conversion from ‘char’ to non-scalar type ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’} requested
<stdin>:4:23: note: in expansion of macro ‘TEST’
<<<OUTPUT ENDS>>>

(executable check - failure) ccls
(executable check - failure) clang-check
(executable check - failure) clangd
(executable check - failure) clang-tidy
(executable check - failure) clazy-standalone
(executable check - failure) cppcheck
(executable check - failure) cpplint
(executable check - failure) cquery
(executable check - failure) flawfinder
```

**Comment by w0rp** [comment]
Could you create a pull request for this with an update to tests? You can include more samples lines of output in the tests, and ensure that changing the regular expression doesn't break anything.

**Comment by Kimplul** [comment]
Will do, but quick question: Since these kinds of errors don't report a line number, just a note where the error occurred, should I leave ```lnum``` as zero or try to make the linter use the note's line number?

One issue with using the note's lnum is that in some cases where the error occurred in some header file, gcc doesn't use the standard ```In file included from: ...```, and ```ParseProblemsInHeaders``` doesn't trigger.

**Comment by w0rp** [comment]
You could look for specifically errors about macro expansions with a separate pattern, and then parse the line above to get an error message.

```
<command-line>: error: conversion from ‘char’ to non-scalar type ‘std::string’ {aka ‘std::__cxx
11::basic_string<char>’} requested
<stdin>:4:23: note: in expansion of macro ‘TEST’
```

You could read the line and column number from the macro expansion `note`, then look for a `warning:` or `error:` line above without a line number and include that in the message.

**Comment by w0rp** [comment]
Fixed in #3327.
